### DISCUSSION TITLE
Color highlight correct prefixes to help avoid typos early

### DISCUSSION THREAD

**Discussion by eugenesvk** [xonsh/xonsh#5653] [Ideas]
Currently color highlighting is only working on full matches, so
- `python` is highlighted
- `py` isn't (let's ignore the fact that `py` exists and assume we have only a single `python` binary)

But it would be great to catch typos early: test whether any typed text is part of an existing binary and highlight it (not in green like the full binary match, but with another "partial match" color), so:
- `py` is partial-highlighted
- `pyth` is
- `pyht` is NOT, so you can see that you've made a typo before you need to type the full `pyhton` and see no green

Something like this where valid partial commands are underlined

https://github.com/user-attachments/assets/f6373b33-eb29-43de-b1af-f43f4acf3b5a

**Comment by anki-code** [comment]
I think the good way to implement this feature is creating distinct pluggable [xontrib](https://github.com/topics/xontrib) (see also [xontrib template](https://github.com/xonsh/xontrib-template)). We have highlighting as [pygments hook](https://github.com/xonsh/xonsh/blob/938db85a8733eb15be76d8b599c0d6dfc68b7cac/xonsh/pyghooks.py#L1595-L1599). It will be cool new wave to have a way to make it pluggable and extensible without any changes in xonsh core.

**Reply by eugenesvk** [reply]
pygments hook is part of the core, how do you replace it with a xontrib (the template doesn't have this info)? Or did you mean remove pygments hook from core altogether to a xontrib?

**Reply by anki-code** [reply]
Xonsh long term intention is to have an ability to extend any part of the shell without updating the core code so we have xontribs, completers, events, decorators, etc. If we can create "hook for pygments hook" like event as example or something like this it will be interesting to see how it can work.

**Reply by eugenesvk** [reply]
so far added a simpler way (without the complicated new event system) and only for the dir caches in https://github.com/xonsh/xonsh/pull/5660
